### ISSUE TITLE
Page bundling

### ISSUE THREAD

**Issue by pfraze** [ssbc/phoenix#401]
In #400 we've begun exploring HTML publishing and navigating on the blockchain. HTML is a good first asset to explore, because it's a wrapper format: it can import other assets into itself, and embed styling and code, so it can accomplish a lot of tasks.

However, there are still some downsides with using HTML exclusively:

 - **It does a mediocre job coordinating with other blobs.** Without a namespace indirection, all assets have to be imported by hash. This means manually editing references in the HTML each time an asset version is modified.
 - **It is not capable of maintaining metadata across versions.** Subsequent publishings of HTML changes the hash-id. In order to maintain a relationship with past versions or forks, Patchwork needs to track past hash-ids, or it needs to prompt the user to fill them in manually during publishing. (The latter could be made less painful by having the user go to the parent version of a file, and having them click "Publish Fork" or "Publish New Version," then select the new file from their HD.)
 - **Lack of meta-identity complicates referencing from other interfaces.** For instance, an issue tracker will be much more effective when it can persist across versions.

These issues can be solved with a "bundle" message type, which references the assets a page needs. The `bundle` would include a shortname map for the assets, allowing lookup from the HTML without resorting to a hash-id.

Because of the lookup's indirection, an HTML page will not have a fixed behavior for a given hash. This is because imported assets could be resolved differently from one bundle to another. Only the `bundle` message will have a fixed behavior for a given ID. Therefore, Patchwork should only execute `bundles`, not HTML files. HTML files should instead be rendered in plaintext. (We can add a navigation sidepanel, to render backlinks to its `bundles`.)

Bundle meta-identity can work in a couple ways:

 - We can establish a namespace for the publishing user, and let each new `bundle` claim its name on publish. This would create a global lookup for the package under the publishing user's ID, which could be used to create relatively URL-like IDs, such as `pfraze/suggest-box`. Unfortunately, these IDs will not be universal without a globally-unique username mapping. Therefore, the advantage of this approach is dubious: we'd need to use `$PUBKEY/suggest-box` to have URL behavior.
 - We can use the hash ID of the first version of a bundle as its persistent identifier. This would result in a tree-structure of `bundle` versions. The user would be able to navigate this tree, and developers could publish metadata about the versions, including commit messages and branch names

An interesting quality of the latter option is, if we let users select a preferred version or branch, then opening the `bundle` by meta ID would involve an indirection, and thus differ between devices. This would make the meta-ID act like a URL, so long as the various versions remain compatible, but also allow per-device customization.

**Comment by pfraze** [comment]
It will be useful to store metadata (such as asset-listing) in local files, so that development versions of the `bundle` can be persisted. Development versions should also be located somewhere in the file-system, chosen by the developer, in order to let external editors work on the assets.

If we do not restrict the user to having one copy of a `bundle` on the disk at a time, then we'll need to be ready for multiple active "checkouts."

**Comment by dominictarr** [comment]
yes. i'm thinking this could be general purpose, like you could refer to a portion of a bundle using `%{bundle}/key` and then if an update to that changes .key then you'd get that.

**Comment by dominictarr** [comment]
hmm, and you could refer to an exact version in a bundle by giving the hash the last message update?

**Comment by pfraze** [comment]
If we did the `{bundle}/key` method you describe, the format could be very straightforward! Each new message would overwrite the keys of the last one. We'd just need some special keys to link to parent versions. 

Example sequence:

```js
{
  type: 'bundle', // special
  'index.html': { link: BlobRef, type: 'text/html', size: 500 },
  'pic1.png': { link: BlobRef, type: 'image/png', size: 1000, width: 300, height: 200 },
  'pic2.png': { link: BlobRef, type: 'image/png', size: 1000, width: 300, height: 200 },
  'logo.gif': { link: BlobRef, type: 'image/gif', size: 30, width: 5, height: 5 }
}
{
  type: 'bundle', // special
  root: MsgLink, // special, the above msg
  branch: MsgLink, // special, the above msg
  'index.html': { link: BlobRef, type: 'text/html', size: 600 }, // updated file
  'pic3.png': { link: BlobRef, type: 'image/png', size: 1000, width: 300, height: 200 }, // new pic
  'logo.gif': false // removed
}
```

In the update message, if a key is unspecified, it is unchanged from the previous version. Here's the final bundle:

```js
{
  'index.html': { link: BlobRef, type: 'text/html', size: 600 },
  'pic1.png': { link: BlobRef, type: 'image/png', size: 1000, width: 300, height: 200 },
  'pic2.png': { link: BlobRef, type: 'image/png', size: 1000, width: 300, height: 200 },
  'pic3.png': { link: BlobRef, type: 'image/png', size: 1000, width: 300, height: 200 }
}

**Comment by pfraze** [comment]
Other thoughts:

 - Obviously, sub-objects are sub-directories. `$msg/foo/bar == { foo: { bar: BlobLink } }`
 - Might be smart to put a bundle-type on the message. Something to say, "this bundle is a web app" vs "this bundle is a JS module" vs "this bundle is a photo album."
   - Option 1: we use a different `type` for different bundles. That would mean the bundle is a pattern, not a message-type. It's a similar issue to what we discussed in https://github.com/ssbc/ssb-msg-schemas/issues/9.
   - Option 2: we use `type: bundle` and something like `bundleType: X`. This would mean all bundles of all types can be queried at once, then subdivided -- possibly a good thing.
